README.md
# Simple-encrypt

This is a simple, tiny encrypt package for encrypt your record in db through aes 128, 192, 256 cbc depends on the length of the key.

It can be used to encrypt string and int.

# Install

go get -u github.com/lufishggg/simple-encrypt

# Usage

If we already have a table 'users' like this (use pg as example), and there have been some records:

    CREATE TABLE users (
      id BIGSERIAL PRIMARY KEY,
      name VARCHAR,
      email VARCHAR UNIQUE,
      gender smallint DEFAULT 0,
      created_at TIMESTAMP NOT NULL,
      updated_at TIMESTAMP NOT NULL
    )
	
	id |   name  |      email       |        gender      |          created_at           |             updated_at
    ------------------------------------------------------------------------------------------------------------------------
    1  |  john	 | john@gmail.com	|          1	     |   2019-11-05 21:09:20.30628	 |      2019-11-05 21:09:20.30628
    2  |  mike	 | mike@gmail.com	|          2	     |   2019-11-05 21:09:20.30628	 |      2019-11-05 21:09:20.30628
    3  |  Tom	 |	                |          1	     |   2019-11-05 21:09:20.30628	 |      2019-11-05 21:09:20.30628
    4  |  ""	 |	                |          0	     |   2019-11-05 21:09:20.30628	 |      2019-11-05 21:09:20.30628
    
Now we want to encrypt the remain records (even gender!) and future records. Just simpley scan the value and update again.

First, open the db:

    db, _ := sqlx.Open("postgres", "postgres://example:example@127.0.0.1:5432/example?sslmode=disable")
    defer db.Close()
    
Then, create a new column call encrypted_gender to encrypt the gender:

   	_, _ = db.Exec("ALTER TABLE users ADD encrypted_gender VARCHAR DEFAULT '0';")
   	
Then, init the key from config file or environment variables or any secure way you like, don't do as the example:
    
   	_ = se.InitDefaultKey("0123456789abcdef0123456789abcdef")

That is all we should do. Then just do db operation you like:

    import (
        "fmt"
        "github.com/jmoiron/sqlx"
        _ "github.com/lib/pq"
        se "github.com/lufishggg/simple-encrypt"
        "time"
        "xorm.io/builder"
    )
    
    type User struct {
    	Id              int64            `db:"id"`
    	Name            se.EncryptString `db:"name"`
    	Email           se.EncryptString `db:"email"`
    	Gender          int              `db:"gender"`
    	EncryptedGender se.EncryptInt    `db:"encrypted_gender"`
    	CreatedAt       time.Time        `db:"created_at"`
    	UpdatedAt       time.Time        `db:"updated_at"`
    }
    
    func main() {
    	db, _ := sqlx.Open("postgres", "postgres://example:example@127.0.0.1:5432/example?sslmode=disable")
    	defer db.Close()
    	
    	// First we have to create a new column for encrypting gender, let's call it encrypted_gender
    	_, _ = db.Exec("ALTER TABLE users ADD encrypted_gender VARCHAR DEFAULT '0';")
    	
    	// We have to init the key. That is all what we should do!
    	_ = se.InitDefaultKey("0123456789abcdef0123456789abcdef")
    	
        user := User{
            Name:            se.NewEncryptString(nil),
            Email:           se.NewEncryptString(nil),
            EncryptedGender: se.NewEncryptInt(nil),
        }
        
        for i := 1; i < 5; i++ {
            // Get original records
            query, args, _ := builder.Postgres().Select("*").From("users").Where(builder.Eq{"id": i}).ToSQL()
            
            // Attention that even if there is some error (it is unavoidable because the original records are not cipher texts), we still scan the values!
            // This is for encrypting the remain data that we can easily encrypt the data just by simply scan and insert
            _ = db.Get(&user, query, args...)
            
            fmt.Println("original data in database:")
            fmt.Printf("id: %d, name: %s, email: %s, gender: %d, encrypted_gender: %d\n", user.Id, user.Name.String(), user.Email.String(), user.Gender, user.EncryptedGender.Int())
            
            // Simply update the data in the table again
            query, args, _ = builder.Postgres().Update(builder.Eq{
                "name":             user.Name,
                "email":            user.Email,
                "gender":           0,                              // Set gender to be 0
                "encrypted_gender": se.NewEncryptInt(&user.Gender), // Set encrypted_gender to be original gender
            }).From("users").Where(builder.Eq{"id": i}).ToSQL()
            _, _ = db.Exec(query, args...)
            
            // Get encrypted data
            query, args, _ = builder.Postgres().Select("*").From("users").Where(builder.Eq{"id": i}).ToSQL()
            _ = db.Get(&user, query, args...)
            fmt.Println("encrypted data in database, it is absolutely the same with the original data!:")
            fmt.Printf("id: %d, name: %s, email: %s, gender: %d, encrypted_gender: %d\n", user.Id, user.Name.String(), user.Email.String(), user.Gender, user.EncryptedGender.Int())
        }
	}
	
Here is what we get from output:

    original data in database:
	id: 1, name: john, email: john@gmail.com, gender: 1, encrypted_gender: 0

	encrypted data in database, it is absolutely the same with the original data!:
	id: 1, name: john, email: john@gmail.com, gender: 0, encrypted_gender: 1

	original data in database:
	id: 2, name: mike, email: mike@gmail.com, gender: 2, encrypted_gender: 0

	encrypted data in database, it is absolutely the same with the original data!:
	id: 2, name: mike, email: mike@gmail.com, gender: 0, encrypted_gender: 2

	original data in database:
	id: 3, name: tom, email: , gender: 1, encrypted_gender: 0

	encrypted data in database, it is absolutely the same with the original data!:
	id: 3, name: tom, email: , gender: 0, encrypted_gender: 1

	original data in database:
	id: 4, name: "", email: , gender: 0, encrypted_gender: 0

	encrypted data in database, it is absolutely the same with the original data!:
	id: 4, name: "", email: , gender: 0, encrypted_gender: 0
	
But actually, the database is new like:

		id     |                          name                     |                  email                       | gender |         created_at            |         updated_at          | 		       encrypted_gender
	---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
		1      |      p7XdzCkfo4V4HeX/ejSC48t315BZaJkV182aikZK/Qw= | BfjZss+4SgoaxPWKR8kKmzfMB9VVx3qy9gZ4R1ep1hA= |    0   |    2019-11-05 21:09:20.30628  |  2019-11-05 21:09:20.30628  |   HRAgIgXmlSqZ5bs483ekL4fahD7dhexlS/Fs9Ezr/Ec=
		2      |      wtifiMEwTRXXw5pAbfVlkzSpBzJXTzg/Xq9kVsy3CaE= | WaxG+BZ/UKHwJ7+6qAcKUaqccPTRUdg15A9oTR0vVw4= |    0   |    2019-11-05 21:09:20.30628  |  2019-11-05 21:09:20.30628  |   i6T6BuzdcWZKHBGoeJX1KU667FuK4NR6IHhsepArWx0=
		3      |      lPJLPkbNREZEBkmAL62OkgJk806xIZn5IVyhu2/patk= |	                                          |    0   |    2019-11-05 21:09:20.30628  |  2019-11-05 21:09:20.30628  |   jzAf9hN14eLoM5zYyeHJ9iTNi/4siEMB6IQFmXQ4oRg=
		4      |      3foz2J7IG7BNn29IOUUsNGwRDPX2q9R6ezy2YlNFoXE= |	                                       	  |    0   |    2019-11-05 21:09:20.30628  |  2019-11-05 21:09:20.30628  |   A19kMvFUJbSAQNYI4HykbTvQe/EhEyuozEOndZwJgZw=

CHANGELOG.md
# Changelog

## v1.1
Moves example to README.md.

## v1.0
Basic encrypt string and int data into database.

simple_encrypt_test.go
package simple_encrypt

import (
	"crypto/aes"
	"encoding/hex"
	"github.com/stretchr/testify/assert"
	"testing"
)

// TestInitDefaultKey tests the correctness of the default key initialization
func TestInitDefaultKey(t *testing.T) {
	assert.Error(t, InitDefaultKey("0123456789"), "key length should be 32, 48 or 64")
	assert.Error(t, InitDefaultKey("abcdef0123456789abcdefghijklmnop"), "key should be hex string")
	_ = InitDefaultKey("0123456789abcdef0123456789abcdef")
	block, _ := initBlock("0123456789abcdef0123456789abcdef")
	assert.Equal(t, block, keys[DefaultKeyName], "should be the right default key")
	assert.Error(t, InitDefaultKey("0123456789abcdef0123456789abcdef"), "should not init twice")
}

// TestInitKeys tests the correctness of the keys initialization
func TestInitKeys(t *testing.T) {
	_ = InitKeys(map[string]string{"test1": "0123456789abcdef0123456789abcdef"})
	block, _ := initBlock("0123456789abcdef0123456789abcdef")
	assert.Equal(t, block, keys["test1"], "should be the right test1 key")
	assert.Error(t, InitKeys(map[string]string{"test1": "0123456789abcdef0123456789abcdef"}), "should not init twice")
}

// TestEncryptAndDecrypt tests the correctness of the encrypt and decrypt function
func TestEncryptAndDecrypt(t *testing.T) {
	keyBytes, _ := hex.DecodeString("0123456789abcdef0123456789abcdef")
	block, _ := aes.NewCipher(keyBytes)
	plainTextString := "I am plain text"
	cipherTextString, _ := encrypt(plainTextString, &block)
	plainTextStringNew, _ := decrypt(cipherTextString, &block)
	assert.Equal(t, "I am plain text", plainTextStringNew, "should be the same")
}

// TestEncryptStringScan1 tests the correctness of the scan function
func TestEncryptStringScan1(t *testing.T) {
	es := NewEncryptStringWithKeyName("test", nil)
	_ = InitKeys(map[string]string{"test": "0123456789abcdef0123456789abcdef"})
	_ = es.Scan("BFpTEtSThsjeCnt7wDbTLGFBmbdgGHATtsDb5Fty9Rs=")
	assert.Equal(t, "I am plain text", es.String(), "should get the right plain text")
}

// TestEncryptStringScan2 tests the correctness of the scan function
func TestEncryptStringScan2(t *testing.T) {
	es := NewEncryptStringWithKeyName("test", nil)
	_ = InitKeys(map[string]string{"test": "0123456789abcdef0123456789abcdef"})
	_ = es.Scan(nil)
	assert.Equal(t, true, es.raw == nil, "should get nil if src is nil")
	_ = es.Scan("abc")
	assert.Equal(t, "abc", es.String(), "should return original data if it is not a cipher text")
}

// TestEncryptStringValue tests the correctness of the value function
func TestEncryptStringValue(t *testing.T) {
	raw := "I am plain text"
	es := NewEncryptStringWithKeyName("test", &raw)
	_ = InitKeys(map[string]string{"test": "0123456789abcdef0123456789abcdef"})
	value, _ := es.Value()
	cipherTextString := value.(string)
	_ = es.Scan(cipherTextString)
	assert.Equal(t, "I am plain text", es.String(), "should get the right plain text")
}

// TestEncryptStringInt tests the correctness of the scan function
func TestEncryptIntScan(t *testing.T) {
	ei := NewEncryptIntWithKeyName("test", nil)
	_ = InitKeys(map[string]string{"test": "0123456789abcdef0123456789abcdef"})
	_ = ei.Scan("mBYP/b2SJ/uQjUjW16FWT4xNvLS2a1Q+gBc8Y8pc8TM=")
	assert.Equal(t, 12345, ei.Int(), "should get the right plain int")
}

// TestEncryptStringValue tests the correctness of the value function
func TestEncryptIntValue(t *testing.T) {
	raw := 12345
	ei := NewEncryptIntWithKeyName("test", &raw)
	_ = InitKeys(map[string]string{"test": "0123456789abcdef0123456789abcdef"})
	value, _ := ei.Value()
	cipherTextString := value.(string)
	_ = ei.Scan(cipherTextString)
	assert.Equal(t, 12345, ei.Int(), "should get the right plain int")
}


key.go
package simple_encrypt

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/hex"
	"errors"
)

// To save the keys. You can simply use default key,
// or you can init multiple keys, and encrypt different record with different keys
var keys map[string]*cipher.Block

const DefaultKeyName = "default"

func init() {
	keys = make(map[string]*cipher.Block, 0)
}

// InitDefaultKey inits the default key
func InitDefaultKey(key string) error {
	if keys[DefaultKeyName] != nil {
		return errors.New("do not init the key twice")
	}
	block, err := initBlock(key)
	if err != nil {
		return err
	}
	keys[DefaultKeyName] = block
	return nil
}

// InitKeys inits multiple keys with unique names to encrypts records with different keys
func InitKeys(keysMap map[string]string) error {
	// Check if there are duplicate keys and illegal keys
	blocks := make(map[string]*cipher.Block, 0)
	for name, key := range keysMap {
		if keys[name] != nil {
			return errors.New("do not init the key twice")
		}
		block, err := initBlock(key)
		if err != nil {
			return err
		}
		blocks[name] = block
	}
	for name, block := range blocks {
		keys[name] = block
	}
	return nil
}

// initBlock inits a block from given key
func initBlock(key string) (*cipher.Block, error) {
	// key must be a hex string with length of 32, 48 or 64, corresponding to aes 128, 192, 256 cbc
	keyBytes, err := hex.DecodeString(key)
	if err != nil {
		return nil, err
	}
	block, err := aes.NewCipher(keyBytes)
	if err != nil {
		return nil, err
	}
	return &block, nil
}


decrypt.go
package simple_encrypt

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"
	"errors"
)

// decrypt decrypts the data through aes
func decrypt(cipherTextString string, block *cipher.Block) (string, error) {
	if block == nil {
		return "", errors.New("block is not initialized")
	}
	cipherTextBytes, err := base64.StdEncoding.DecodeString(cipherTextString)
	if err != nil {
		return "", err
	}
	if len(cipherTextBytes) < aes.BlockSize {
		return "", errors.New("cipher text is too short")
	}
	// Get the iv from cipher text
	iv := cipherTextBytes[:aes.BlockSize]
	cipherTextBytes = cipherTextBytes[aes.BlockSize:]
	if len(cipherTextBytes)%aes.BlockSize != 0 {
		return "", errors.New("cipher text is not a multiple of the block size")
	}
	// Decrypt
	mode := cipher.NewCBCDecrypter(*block, iv)
	mode.CryptBlocks(cipherTextBytes, cipherTextBytes)
	// Unpadding
	cipherTextBytes = pkcs7Unpadding(cipherTextBytes)
	ciphertextStr := string(cipherTextBytes)
	return ciphertextStr, nil
}


encrypt.go
package simple_encrypt

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"errors"
	"io"
)

// encrypt encrypts the data through aes
func encrypt(plainTextString string, block *cipher.Block) (string, error) {
	// If block is null
	if block == nil {
		return "", errors.New("haven't init block")
	}
	// Get the plain text bytes from plain text string and padding it
	plainTextBytes := []byte(plainTextString)
	plainTextBytes = pkcs7Padding(plainTextBytes)
	if len(plainTextBytes)%aes.BlockSize != 0 {
		return "", errors.New("plaintext is not a multiple of the block size")
	}
	// Make the cipher text bytes
	cipherTextBytes := make([]byte, aes.BlockSize+len(plainTextBytes))
	// Random generate iv
	iv := cipherTextBytes[:aes.BlockSize]
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return "", err
	}
	// Encrypt
	mode := cipher.NewCBCEncrypter(*block, iv)
	mode.CryptBlocks(cipherTextBytes[aes.BlockSize:], plainTextBytes)
	// Base64 encodes the cipher text
	cipherTextString := base64.StdEncoding.EncodeToString(cipherTextBytes)
	return cipherTextString, nil
}


encrypt_int.go
package simple_encrypt

import (
	"database/sql/driver"
	"errors"
	"strconv"
)

// EncryptInt is for encrypt an int (such as gender, enum). If anything error in the process of scanner or valuer,
// they will return the raw data, instead of nil
type EncryptInt struct {
	keyName string
	raw     *int
}

// Scan scans the value src from the database to the type EncryptInt
func (i *EncryptInt) Scan(src interface{}) error {
	// src might be null
	if src == nil {
		i.raw = nil
		return nil
	}
	// The src type must be string or []byte
	var source string
	switch src.(type) {
	case string:
		source = src.(string)
	case []byte:
		source = string(src.([]byte))
	default:
		i.raw = nil
		return errors.New("incompatible type for scan")
	}
	// Have to recover from panic because we always want to get the original data no matter what happen
	defer recoverFromPanicEncryptInt(i, source)
	// Decrypt the source by key with given key name
	plainTextString, err := decrypt(source, keys[i.keyName])
	if err != nil {
		tmp, err2 := strconv.Atoi(source)
		if err2 != nil {
			i.raw = nil
			return err2
		}
		i.raw = &tmp
		return err
	}
	tmp, err := strconv.Atoi(plainTextString)
	if err != nil {
		i.raw = nil
		return err
	}
	i.raw = &tmp
	return nil
}

// Value encrypts the value and save in the database
func (i EncryptInt) Value() (driver.Value, error) {
	if i.raw == nil {
		return nil, nil
	}
	plainTextString := strconv.Itoa(*i.raw)
	cipherTextString, err := encrypt(plainTextString, keys[i.keyName])
	if err != nil {
		return plainTextString, err
	}
	return cipherTextString, nil
}

// Int gets the raw int
func (i *EncryptInt) Int() int {
	if i.raw == nil {
		return 0
	}
	return *i.raw
}

// SetKeyName sets the key
func (i *EncryptInt) SetKeyName(keyName string) {
	i.keyName = keyName
}

// SetRaw sets the raw
func (i *EncryptInt) SetRaw(raw *int)  {
	i.raw = raw
}

// NewDefaultEncryptInt news an EncryptInt with default key name
// you must only use NewDefaultEncryptInt or NewEncryptInt to new an EncryptInt
func NewEncryptInt(raw *int) EncryptInt {
	return EncryptInt{
		keyName: DefaultKeyName,
		raw:     raw,
	}
}

// NewEncryptInt news an EncryptInt with given key name
// you must only use NewDefaultEncryptInt or NewEncryptInt to new an EncryptInt
func NewEncryptIntWithKeyName(keyName string, raw *int) EncryptInt {
	return EncryptInt{
		keyName: keyName,
		raw:     raw,
	}
}

// recoverFromPanicEncryptInt recovers from panic and get the plain int for EncryptInt
func recoverFromPanicEncryptInt(i *EncryptInt, source string) {
	if r := recover(); r != nil {
		tmp, err := strconv.Atoi(source)
		if err != nil {
			i.raw = nil
		}
		i.raw = &tmp
	}
}


encrypt_string.go
package simple_encrypt

import (
	"database/sql/driver"
	"errors"
)

// EncryptString is for encrypt a string. If anything error in the process of scanner or valuer,
// they will return the raw data, instead of nil
type EncryptString struct {
	keyName string
	raw     *string
}

// Scan scans the value src from the database to the type EncryptString
func (s *EncryptString) Scan(src interface{}) error {
	// src might be null
	if src == nil {
		s.raw = nil
		return nil
	}
	// The src type must be string or []byte
	var source string
	switch src.(type) {
	case string:
		source = src.(string)
	case []byte:
		source = string(src.([]byte))
	default:
		s.raw = nil
		return errors.New("incompatible type for scan")
	}
	// Have to recover from panic because we always want to get the original data no matter what happen
	defer recoverFromPanicEncryptString(s, source)
	// Decrypt the source by key with given key name
	plainTextString, err := decrypt(source, keys[s.keyName])
	if err != nil {
		s.raw = &source
		return nil
	}
	s.raw = &plainTextString
	return nil
}

// Value encrypts the value and save in the database
func (s EncryptString) Value() (driver.Value, error) {
	if s.raw == nil {
		return nil, nil
	}
	cipherTextString, err := encrypt(*s.raw, keys[s.keyName])
	if err != nil {
		return *s.raw, err
	}
	return cipherTextString, nil
}

// Int gets the raw string
func (s *EncryptString) String() string {
	if s.raw == nil {
		return ""
	}
	return *s.raw
}

// SetKeyName sets the key
func (s *EncryptString) SetKeyName(keyName string) {
	s.keyName = keyName
}

// SetRaw sets the raw
func (s *EncryptString) SetRaw(raw *string)  {
	s.raw = raw
}

// NewEncryptString news a default EncryptString
func NewEncryptString(raw *string) EncryptString {
	return EncryptString{
		keyName: DefaultKeyName,
		raw:     raw,
	}
}

// NewEncryptStringWithKeyName news an EncryptString with given key name
// you must only use NewDefaultEncryptString or NewEncryptString to new an EncryptString
func NewEncryptStringWithKeyName(keyName string, raw *string) EncryptString {
	return EncryptString{
		keyName: keyName,
		raw:     raw,
	}
}

// recoverFromPanicEncryptString recovers from panic and get the plain string for EncryptString
func recoverFromPanicEncryptString(i *EncryptString, source string) {
	if r := recover(); r != nil {
		i.raw = &source
	}
}


example/main.go
// examples
package main

import (
	"fmt"
	"github.com/jmoiron/sqlx"
	_ "github.com/lib/pq"
	se "github.com/lufishggg/simple-encrypt"
	"time"
	"xorm.io/builder"
)

type User struct {
	Id int64 `db:"id"`
	Name se.EncryptString `db:"name"`
	Email se.EncryptString `db:"email"`
	Gender int `db:"gender"`
	EncryptedGender se.EncryptInt `db:"encrypted_gender"`
	CreatedAt time.Time `db:"created_at"`
	UpdatedAt time.Time `db:"updated_at"`
}

/*
	If we already have a table 'users' like this (use pg as example), and there have been some records:

	CREATE TABLE users (
	  id BIGSERIAL PRIMARY KEY,
	  name VARCHAR,
	  email VARCHAR UNIQUE,
	  gender smallint DEFAULT 0,
	  created_at TIMESTAMP NOT NULL,
	  updated_at TIMESTAMP NOT NULL
	)

	id |   name  |      email       |        gender      |          created_at           |             updated_at
	------------------------------------------------------------------------------------------------------------------------
	1  |  john	 | john@gmail.com	|          1	     |   2019-11-05 21:09:20.30628	 |      2019-11-05 21:09:20.30628
	2  |  mike	 | mike@gmail.com	|          2	     |   2019-11-05 21:09:20.30628	 |      2019-11-05 21:09:20.30628
	3  |  Tom	 |	                |          1	     |   2019-11-05 21:09:20.30628	 |      2019-11-05 21:09:20.30628
	4  |  ""	 |	                |          0	     |   2019-11-05 21:09:20.30628	 |      2019-11-05 21:09:20.30628

	Now we want to encrypt the remain records (even gender!) and future records.
	Run main(), and we can get the output:
		original data in database:
		id: 1, name: john, email: john@gmail.com, gender: 1, encrypted_gender: 0

		encrypted data in database, it is absolutely the same with the original data!:
		id: 1, name: john, email: john@gmail.com, gender: 0, encrypted_gender: 1

		original data in database:
		id: 2, name: mike, email: mike@gmail.com, gender: 2, encrypted_gender: 0

		encrypted data in database, it is absolutely the same with the original data!:
		id: 2, name: mike, email: mike@gmail.com, gender: 0, encrypted_gender: 2

		original data in database:
		id: 3, name: tom, email: , gender: 1, encrypted_gender: 0

		encrypted data in database, it is absolutely the same with the original data!:
		id: 3, name: tom, email: , gender: 0, encrypted_gender: 1

		original data in database:
		id: 4, name: "", email: , gender: 0, encrypted_gender: 0

		encrypted data in database, it is absolutely the same with the original data!:
		id: 4, name: "", email: , gender: 0, encrypted_gender: 0
*/
func main()  {
	db, _ := sqlx.Open("postgres", "postgres://docker:docker@127.0.0.1:35432/gis?sslmode=disable")
	defer db.Close()
	// First we have to create a new column for encrypting gender, let's call it encrypted_gender
	_, _ = db.Exec("ALTER TABLE users ADD encrypted_gender VARCHAR DEFAULT '0';")
	// We have to init the key. That is all what we should do!
	_ = se.InitDefaultKey("0123456789abcdef0123456789abcdef")
	user := User{
		Name:            se.NewEncryptString(nil),
		Email:           se.NewEncryptString(nil),
		EncryptedGender: se.NewEncryptInt(nil),
	}
	for i := 1; i < 5; i++ {
		// Get original records
		query, args, _ := builder.Postgres().Select("*").From("users").Where(builder.Eq{"id": i}).ToSQL()
		// Attention that even if there is some error (it is unavoidable because the original records are not cipher texts), we still scan the values!
		// This is for encrypting the remain data that we can easily encrypt the data just by simply scan and insert
		_ = db.Get(&user, query, args...)
		fmt.Println("original data in database:")
		fmt.Printf("id: %d, name: %s, email: %s, gender: %d, encrypted_gender: %d\n", user.Id, user.Name.String(), user.Email.String(), user.Gender, user.EncryptedGender.Int())
		// Simply update the data in the table again
		query, args, _ = builder.Postgres().Update(builder.Eq{
			"name": user.Name,
			"email": user.Email,
			"gender": 0,// Set gender to be 0
			"encrypted_gender": se.NewEncryptInt(&user.Gender),// Set encrypted_gender to be original gender
		}).From("users").Where(builder.Eq{"id": i}).ToSQL()
		_, _ = db.Exec(query, args...)
		/*
			Here, we have encrypted the data, in database looks like just as:
			id     |                          name                     |                  email                       | gender |         created_at            |         updated_at          | 		       encrypted_gender
		---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
			1      |      p7XdzCkfo4V4HeX/ejSC48t315BZaJkV182aikZK/Qw= | BfjZss+4SgoaxPWKR8kKmzfMB9VVx3qy9gZ4R1ep1hA= |    0   |    2019-11-05 21:09:20.30628  |  2019-11-05 21:09:20.30628  |   HRAgIgXmlSqZ5bs483ekL4fahD7dhexlS/Fs9Ezr/Ec=
			2	   |      wtifiMEwTRXXw5pAbfVlkzSpBzJXTzg/Xq9kVsy3CaE= | WaxG+BZ/UKHwJ7+6qAcKUaqccPTRUdg15A9oTR0vVw4= |	   0   |    2019-11-05 21:09:20.30628  |  2019-11-05 21:09:20.30628  |   i6T6BuzdcWZKHBGoeJX1KU667FuK4NR6IHhsepArWx0=
			3	   |      lPJLPkbNREZEBkmAL62OkgJk806xIZn5IVyhu2/patk= |	                                          |    0   |    2019-11-05 21:09:20.30628  |  2019-11-05 21:09:20.30628  |   jzAf9hN14eLoM5zYyeHJ9iTNi/4siEMB6IQFmXQ4oRg=
			4	   |      3foz2J7IG7BNn29IOUUsNGwRDPX2q9R6ezy2YlNFoXE= |	                                       	  |	   0   |    2019-11-05 21:09:20.30628  |  2019-11-05 21:09:20.30628  |  A19kMvFUJbSAQNYI4HykbTvQe/EhEyuozEOndZwJgZw=

			Next, get the record and check its value
		 */

		// Get encrypted data
		query, args, _ = builder.Postgres().Select("*").From("users").Where(builder.Eq{"id": i}).ToSQL()
		_ = db.Get(&user, query, args...)
		fmt.Println("encrypted data in database, it is absolutely the same with the original data!:")
		fmt.Printf("id: %d, name: %s, email: %s, gender: %d, encrypted_gender: %d\n", user.Id, user.Name.String(), user.Email.String(), user.Gender, user.EncryptedGender.Int())
	}
}
